pdf/model/writer.go

```diff
@@ -132,6 +150,131 @@ func NewPdfWriter() PdfWriter {
+		for _, del := range toDelete {
+			toDeleteMap[del] = struct{}{}
+		}
+		objects = make([]PdfObject, 0, len(this.objects))
+		for _, o := range this.objects {
+			if _, ok := toDeleteMap[o]; ok {
+				delete(this.objectsMap, o)
+			} else {
+				objects = append(objects, o)
+			}
+		}
+	}
+	for _, o := range toAdd {
+		objects = append(objects, o)
+		this.objectsMap[o] = true
+	}
+	this.objects = objects
+}
+
+// copyObject creates deep copy of the Pdf object
+// fills the replace table to replace the old object to the copy of object
+func copyObject(obj PdfObject, replaceTable map[PdfObject]PdfObject) PdfObject {
+	if newObj, ok := replaceTable[obj]; ok {
+		return newObj
+	}
+
+	switch t := obj.(type) {
+	case *PdfObjectArray:
+		newObj := &PdfObjectArray{}
+		replaceTable[obj] = newObj
+		for _, val := range t.Elements() {
+			newObj.Append(copyObject(val, replaceTable))
+		}
+		return newObj
+	case *PdfObjectStreams:
+		newObj := &PdfObjectStreams{PdfObjectReference: t.PdfObjectReference}
+		replaceTable[obj] = newObj
+		for _, val := range t.Elements() {
+			newObj.Append(copyObject(val, replaceTable))
+		}
+		return newObj
+	case *PdfObjectStream:
+		newObj := &PdfObjectStream{
+			Stream:             t.Stream,
+			PdfObjectReference: t.PdfObjectReference,
+		}
+		replaceTable[obj] = newObj
+		newObj.PdfObjectDictionary = copyObject(t.PdfObjectDictionary, replaceTable).(*PdfObjectDictionary)
+		return newObj
+	case *PdfObjectDictionary:
+		newObj := MakeDict()
+		replaceTable[obj] = newObj
+		for _, key := range t.Keys() {
+			val := t.Get(key)
+			newObj.Set(key, copyObject(val, replaceTable))
+		}
+		return newObj
+	case *PdfIndirectObject:
+		newObj := &PdfIndirectObject{
+			PdfObjectReference: t.PdfObjectReference,
+		}
+		replaceTable[obj] = newObj
+		newObj.PdfObject = copyObject(t.PdfObject, replaceTable)
+		return newObj
+	case *PdfObjectString:
+		newObj := &PdfObjectString{}
+		*newObj = *t
+		replaceTable[obj] = newObj
+		return newObj
+	case *PdfObjectName:
+		newObj := PdfObjectName(*t)
+		replaceTable[obj] = &newObj
+		return &newObj
+	case *PdfObjectNull:
+		newObj := PdfObjectNull{}
+		replaceTable[obj] = &newObj
+		return &newObj
+	case *PdfObjectInteger:
+		newObj := PdfObjectInteger(*t)
+		replaceTable[obj] = &newObj
+		return &newObj
+	case *PdfObjectReference:
+		newObj := PdfObjectReference(*t)
+		replaceTable[obj] = &newObj
+		return &newObj
+	case *PdfObjectFloat:
+		newObj := PdfObjectFloat(*t)
+		replaceTable[obj] = &newObj
+		return &newObj
+	case *PdfObjectBool:
+		newObj := PdfObjectBool(*t)
+		replaceTable[obj] = &newObj
+		return &newObj
+	default:
+		common.Log.Info("TODO(a5i): implement copyObject for %+v", obj)
+	}
+	// return other objects as is
+	return obj
+}
+
+// copyObjects makes objects copy and set as working
+func (this *PdfWriter) copyObjects(replaceTable map[PdfObject]PdfObject) {
+	if replaceTable == nil {
+		replaceTable = make(map[PdfObject]PdfObject)
```

Comment:
Can you briefly describe the replaceTable? Seems to me like used as an endpoint for recursion, is that correct?

Reply:
Yes, You are right it's endpoint for recursion.
I renamed  replaceTable to objectToObjectCopyMap and added a description.